pandas/tseries/offsets.py

```diff
@@ -1160,6 +1162,196 @@ def onOffset(self, dt):
+                    # rollforward in the negative direction so add 1
+                    n += 1
+                elif n == 0:
+                    n = 1
+
+        return self._apply(n, other)
+
+    def _apply(self, n, other):
+        """Handle specific apply logic for child classes"""
+        raise AbstractMethodError(self)
+
+    @apply_index_wraps
+    def apply_index(self, i):
+        # determine how many days away from the 1st of the month we are
+        days_from_start = i.to_perioddelta('M').asi8
+        delta = Timedelta(days=self.day_of_month - 1).value
+
+        # get boolean array for each element before the day_of_month
+        before_day_of_month = days_from_start < delta
+
+        # get boolean array for each element after the day_of_month
+        after_day_of_month = days_from_start > delta
+        return self._apply_index(i, before_day_of_month, after_day_of_month)
+
+    def _apply_index(self, i, before_day_of_month, after_day_of_month):
+        """Handle specific apply_index logic for child classes"""
+        raise AbstractMethodError(self)
+
+
+class SemiMonthEnd(SemiMonthOffset):
+    """
+    Two DateOffset's per month repeating on the last
+    day of the month and day_of_month.
+
+    .. versionadded:: 0.18.2
+
+    Parameters
+    ----------
+    n: int
+    normalize : bool, default False
+    day_of_month: int, {1, 3,...,27}, default 15
+    """
+    _prefix = 'SM'
+    _min_day_of_month = 1
+
+    def onOffset(self, dt):
+        if self.normalize and not _is_normalized(dt):
+            return False
+        _, days_in_month = tslib.monthrange(dt.year, dt.month)
+        return dt.day in (self.day_of_month, days_in_month)
+
+    def _apply(self, n, other):
+        # if other.day is not day_of_month move to day_of_month and update n
+        if other.day < self.day_of_month:
+            other += relativedelta(day=self.day_of_month)
+            if n > 0:
+                n -= 1
+        elif other.day > self.day_of_month:
+            other += relativedelta(day=self.day_of_month)
+            if n == 0:
+                n = 1
+            else:
+                n += 1
+
+        months = n // 2
+        day = 31 if n % 2 else self.day_of_month
+        return other + relativedelta(months=months, day=day)
+
+    def _apply_index(self, i, before_day_of_month, after_day_of_month):
+        n = self.n
+        time = i.to_perioddelta('D')
+        is_month_end = i.is_month_end
+        if n > 0:
+            roll_end = np.where(is_month_end, 1, 0)
+            roll_before = np.where(before_day_of_month, n, n + 1)
+            roll = roll_end + roll_before
+        elif n == 0:
+            roll_after = np.where(after_day_of_month, 2, 0)
+            roll_before = np.where(~after_day_of_month, 1, 0)
+            roll = roll_before + roll_after
+        else:
+            roll = np.where(after_day_of_month, n + 2, n + 1)
+
+        return ((i.to_period('M') + (roll // 2)).to_timestamp() +
```

Comment:
I would like to see this split up a bit, a lot going on here, can you make it more clear?